### ISSUE TITLE
DNS load balancer should randomly respond with ALL IPs - just a different order every time

### ISSUE THREAD

**Issue by mykaul** [scylladb/alternator-load-balancing#14]
Instead of responding with just one IP, it should respond with all, in a different order each time.
This allows the client to do its own randomization on the data, connect to multiple nodes (backup nodes), and whatnot.
(Does require the client to be smart enough about this, go over all IPs returned as a response, etc.!)

**Comment by nyh** [comment]
This idea needs to be tested - if it has any practical advantage over the existing situation. Do clients with existing Java/Python/whatever libraries actually get the full list and not just one? Do they do anything smart with this list (like opening connections to all of them) or just stupidly take the first one in the list?

Maybe even if it doesn't work well in today's SDK we can still propose improvements, but I'm not optimistic - Amazon's AWS returns only a single result for the dynamoDB domain - not a list - so they have no incentive to support this use case.

I added a short mention of this idea in the DNS section of https://docs.google.com/document/d/1twgrs6IM1B10BswMBUNqm7bwu5HCm47LOYE-Hdhuu_8/edit.

**Comment by nyh** [comment]
The same idea was suggested again in a separate discussion by @avikivity. He suggests that we check whether this works, in practice, for real SDKs. "Works" in this context means that the client would pick one of the returned IP addresses at random, while "doesn't work" means the client always picks the first IP address on the list.

Note that we could support both types of clients by returning a **randomly shuffled** list of all nodes - some clients would be able to make use of the whole list and some will just use the first one which is random due to the shuffling - but there is only point in doing that if in practice some common SDKs can use the whole list. It is quite easy to patch the trivial 60-lines-of-Python DNS implementation in dns/dns-loadbalancer.py to do this, and check if it works. I'll do it.

**Comment by mykaul** [comment]
I think this (shuffling) is what DNS does with multiple IPs. It's very basic best practice to randomize them. I'm sure there's some switch to turn this on or off.

**Comment by dkropachev** [comment]
> I think this (shuffling) is what DNS does with multiple IPs. It's very basic best practice to randomize them. I'm sure there's some switch to turn this on or off.

In all aws sdks there two or more abstraction layers on top of basic http and dns resolver, we need to make sure nothing is getting broken on the way.

**Comment by mykaul** [comment]
BTW, /localnodes API - does it random the nodes order? Do the clients receiving the response do?

**Comment by nyh** [comment]
I implemented a modified version of dns-loadbalancer.py called dns-loadbalancer-rr.py which returns the whole list of nodes, randomly cycled, instead of one node. It was easy to check it works as a DNS server, but I'm hesitating if to send this PR and mark it Fixes this issue because it's much more difficult to check that clients can actually use this name server with good results.

Maybe it's enough that "dig" works as expected with my implementation in order to close this, and the real client testing should wait for https://github.com/scylladb/alternator-load-balancing/issues/105?

**Comment by nyh** [comment]
> BTW, /localnodes API - does it random the nodes order? Do the clients receiving the response do?

The "/localnodes" API doesn't randomize the order. Also, we call it rather infrequently (e.g., once per second, once per minute, or whatever) - not for every individual request. Our code can take it and randomize it however it wants, and can do it differently on each request, without refetching /localnodes.

**Comment by dkropachev** [comment]
> I implemented a modified version of dns-loadbalancer.py called dns-loadbalancer-rr.py which returns the whole list of nodes, randomly cycled, instead of one node. It was easy to check it works as a DNS server, but I'm hesitating if to send this PR and mark it Fixes this issue because it's much more difficult to check that clients can actually use this name server with good results.
> 
> Maybe it's enough that "dig" works as expected with my implementation in order to close this, and the real client testing should wait for [#105](https://github.com/scylladb/alternator-load-balancing/issues/105)?

On my book it perfecly find to close this issue when it is merged and do the rest in #105 .

**Comment by avikivity** [comment]
> I implemented a modified version of dns-loadbalancer.py called dns-loadbalancer-rr.py which returns the whole list of nodes, randomly cycled, instead of one node. It was easy to check it works as a DNS server, but I'm hesitating if to send this PR and mark it Fixes this issue because it's much more difficult to check that clients can actually use this name server with good results.
> 
> Maybe it's enough that "dig" works as expected with my implementation in order to close this, and the real client testing should wait for [#105](https://github.com/scylladb/alternator-load-balancing/issues/105)?

We should update a real dns server, not serve dns ourselves.

**Comment by dkropachev** [comment]
> > I implemented a modified version of dns-loadbalancer.py called dns-loadbalancer-rr.py which returns the whole list of nodes, randomly cycled, instead of one node. It was easy to check it works as a DNS server, but I'm hesitating if to send this PR and mark it Fixes this issue because it's much more difficult to check that clients can actually use this name server with good results.
> > Maybe it's enough that "dig" works as expected with my implementation in order to close this, and the real client testing should wait for [#105](https://github.com/scylladb/alternator-load-balancing/issues/105)?
> 
> We should update a real dns server, not serve dns ourselves.

Are you talking about particular deployment, say cloud ?

**Comment by nyh** [comment]
> We should update a real dns server, not serve dns ourselves.

The choice of "real dns server" depends on other things beyond this repository. For example if Scylla Cloud chooses to use "bind" or "route 53" or I don't know what, we'll indeed need to configure it instead of implementing our own name server, but the way to do this will be very different in bind or route 53 or whatever.

Out of the 60 lines of Python that our example implementation has, about 20 lines are to retrieve the list of nodes - and only this part can be copied to code that will create a configuration file, or configuration API calls, or I don't know what - for the "real dns server".

In the mean time, if we want to check if clients can work with a DNS server (as in #105), it's easier to check it with a "non real" DNS server than having the test configure AWS Route 53 or some elaborate "bind" installation or something.

**Comment by avikivity** [comment]
> > > I implemented a modified version of dns-loadbalancer.py called dns-loadbalancer-rr.py which returns the whole list of nodes, randomly cycled, instead of one node. It was easy to check it works as a DNS server, but I'm hesitating if to send this PR and mark it Fixes this issue because it's much more difficult to check that clients can actually use this name server with good results.
> > > Maybe it's enough that "dig" works as expected with my implementation in order to close this, and the real client testing should wait for [#105](https://github.com/scylladb/alternator-load-balancing/issues/105)?
> > 
> > 
> > We should update a real dns server, not serve dns ourselves.
> 
> Are you talking about particular deployment, say cloud ?

Yes

**Comment by avikivity** [comment]
> > We should update a real dns server, not serve dns ourselves.
> 
> The choice of "real dns server" depends on other things beyond this repository. For example if Scylla Cloud chooses to use "bind" or "route 53" or I don't know what, we'll indeed need to configure it instead of implementing our own name server, but the way to do this will be very different in bind or route 53 or whatever.
> 
> Out of the 60 lines of Python that our example implementation has, about 20 lines are to retrieve the list of nodes - and only this part can be copied to code that will create a configuration file, or configuration API calls, or I don't know what - for the "real dns server".
> 
> In the mean time, if we want to check if clients can work with a DNS server (as in [#105](https://github.com/scylladb/alternator-load-balancing/issues/105)), it's easier to check it with a "non real" DNS server than having the test configure AWS Route 53 or some elaborate "bind" installation or something.

Of course, for testing or an example it's fine. And in fact you build on a dns server library.

We could easily let the real dns server recurse into this one.
